network/components/virt_rx.c

```diff
@@ -60,14 +61,42 @@ void rx_return(void)
             assert(!err);
 
             buffer.io_or_offset = buffer.io_or_offset - buffer_data_paddr;
-            microkit_arm_vspace_data_invalidate(buffer.io_or_offset + buffer_data_vaddr, buffer.io_or_offset + buffer_data_vaddr + ROUND_UP(buffer.len, CONFIG_L1_CACHE_LINE_SIZE_BITS));
-
-            int client = get_client((struct ethernet_header *) (buffer.io_or_offset + buffer_data_vaddr));
+            uintptr_t buffer_vaddr = buffer.io_or_offset + buffer_data_vaddr;
+
+            // Cache invalidate after DMA write, so we don't read stale data.
```

Comment:
Thank you for writing this comment, I am certain that it will become useful in the future when we forget that the system call is necessary and user-space can't invalidate.